### PR TITLE
Blogapp user profile

### PR DESCRIPTION
**Pull request by Ali-D-Akbar**
# Blog App Backend:

Description: Custom User Model named `Profile` and it's API endpoints have been added.

Installation Instructions:
1. make a virtual environment by running `virtualenv env`
2. Open terminal and start virtual environment by running `source ./env/bin/activate`
3. run `pip install -r requirements.txt --no-index --find-links file:///tmp/packages` to make requirements
4. run `python manage.py runserver` to start the django server

### CODE REVIEW

File: accounts/admin.py [hunk 1]
```diff
@@ -1 +1,3 @@
 # Register your models here.
+
```
Comment (zubair-arbi) [inline accounts/admin.py]: nit: remove these extra lines

File: blog/models.py [hunk 2]
```diff
@@ -75,3 +79,25 @@ class UserVote(models.Model):
...
+
+
+class Profile(models.Model):
+    user = models.OneToOneField(User, related_name='profile', on_delete=models.CASCADE)
+    GENDER_CHOICES = (
+        ('M', 'Male'),
+        ('F', 'Female'),
+    )
+    gender = models.CharField(max_length=1, choices=GENDER_CHOICES, blank=True)
+    contact_number = models.TextField(max_length=30, blank=True)
```
Comment (zubair-arbi) [inline blog/models.py]: This should be a CharField

File: blog/tests.py [hunk 3]
```diff
@@ -1,7 +1,8 @@
 # Create your tests here.
 from django.contrib.auth.models import User
 from rest_framework import status
-from rest_framework.test import APITestCase, APIRequestFactory, force_authenticate
+from rest_framework.test import (APIRequestFactory, APITestCase,
```
Comment (zubair-arbi) [inline blog/tests.py]: We can improve readability:
```python
from rest_framework.test import (
  APIRequestFactory,
  APITestCase,
  force_authenticate,
)

File: blog/views.py [hunk 4]
```diff
@@ -1,13 +1,15 @@
...
+from rest_framework import permissions, viewsets
 from rest_framework.decorators import action
+from rest_framework.generics import get_object_or_404
 from rest_framework.response import Response
 
-from blog.models import Blog, Comment, UserVote
+from blog.models import Blog, Comment, UserVote, Profile
 from blog.permissions import IsOwnerOrReadOnly
-from blog.serializers import BlogSerializer, UserSerializer, CommentSerializer, VoteSerializer
+from blog.serializers import (BlogSerializer, CommentSerializer,
```
Comment (zubair-arbi) [inline blog/views.py]: Make these group of imports properly indented.

File: webblog/settings/base.py [hunk 5]
```diff
@@ -112,4 +121,29 @@
...
+CORS_ORIGIN_ALLOW_ALL = True
+
+MEDIA_URL = '/media/'
+MEDIA_ROOT = os.path.join(BASE_DIR, "media")
+
+# STATICFILES_DIRS = [
+#     os.path.join(BASE_DIR, './static'),
+# ]
+
+AWS_ACCESS_KEY_ID = config('AWS_ACCESS_KEY_ID')
```
Comment (zubair-arbi) [inline webblog/settings/base.py]: Move this to production.
The uploaded images/file should be stored and served by django locally and for production they should be stored and served from AWS S3.

File: webblog/urls.py [hunk 6]
```diff
@@ -16,13 +16,13 @@
...
 urlpatterns = [
     path('admin/', admin.site.urls),
     path('api/', include('blog.urls')),
     path('', include('accounts.urls')),
 ]
 
-if settings.DEBUG:
-    urlpatterns += static(settings.MEDIA_URL, document_root=settings.MEDIA_ROOT)
+# if settings.DEBUG:
+#     urlpatterns += static(settings.MEDIA_URL, document_root=settings.MEDIA_ROOT)
```
Comment (zubair-arbi) [inline webblog/urls.py]: We need this for local file upload during development.

File: accounts/admin.py (hunk 1 above)
Comment (Ali-D-Akbar) [reply accounts/admin.py]: it's been removed.

File: blog/views.py (hunk 4 above)
Comment (Ali-D-Akbar) [reply blog/views.py]: Done.

File: webblog/settings/base.py (hunk 5 above)
Comment (Ali-D-Akbar) [reply webblog/settings/base.py]: Separated media storage for local and production.

File: webblog/urls.py (hunk 6 above)
Comment (Ali-D-Akbar) [reply webblog/urls.py]: Done.

File: README.md [hunk 7]
```diff
@@ -7,3 +7,4 @@ Installation Instructions:
 2. Open terminal and start virtual environment by running `source ./env/bin/activate`
 3. run `pip install -r requirements.txt --no-index --find-links file:///tmp/packages` to make requirements
 4. run `python manage.py runserver` to start the django server
+
```
Comment (zubair-arbi) [inline README.md]: you haven't rebased this PR after merging your #8

File: blog/admin.py [hunk 8]
```diff
@@ -1,7 +1,58 @@
...
 from django.contrib import admin
+from django.contrib.auth.admin import UserAdmin
+from django.contrib.auth.models import User
+
+
+class ProfileInline(admin.StackedInline):
+    model = Profile
+    can_delete = False
+    verbose_name_plural = 'Profile'
+    fk_name = 'user'
```
Comment (zubair-arbi) [inline blog/admin.py]: nit: replace `fk` with something verbose.

File: blog/admin.py [hunk 9]
```diff
@@ -1,7 +1,58 @@
+from blog.models import Blog, Comment, UserVote, Profile
```
Comment (zubair-arbi) [inline blog/admin.py]: local imports should be after the third party import.

Use isort for ordering the imports automatically. You will have to create a `Makefile` from where you can run your `make quality` command to run the quality.

File: blog/admin.py [hunk 10]
```diff
@@ -1,7 +1,58 @@
...
+    model = Profile
+    can_delete = False
+    verbose_name_plural = 'Profile'
+    fk_name = 'user'
+
+
+class CustomUserAdmin(UserAdmin):
+    inlines = (ProfileInline, )
+    list_display = (
+        'get_profile_id', 'id', 'username', 'first_name', 'last_name', 'get_date_of_birth', 'email', 'get_gender',
```
Comment (zubair-arbi) [inline blog/admin.py]: no need to display the `id`, `get_profile_id`

File: blog/admin.py [hunk 11]
```diff
@@ -1,7 +1,58 @@
...
+        'get_contact_number', 'get_country'
+    )
+    list_select_related = ('profile',)
+
+    def get_date_of_birth(self, instance):
+        return instance.profile.date_of_birth
```
Comment (zubair-arbi) [inline blog/admin.py]: We don't need to add method for direct model relations, e.g. you should be able to see the date of birth by simply adding `profile__date_of_birth` in the `list_display`. 

Same for other field methods here.

File: blog/models.py [hunk 12]
```diff
@@ -75,3 +79,25 @@ class UserVote(models.Model):
 
     class Meta:
         unique_together = ('user', 'blog')
+
+
+class Profile(models.Model):
+    user = models.OneToOneField(User, related_name='profile', on_delete=models.CASCADE)
+    GENDER_CHOICES = (
```
Comment (zubair-arbi) [inline blog/models.py]: Constants should be declared either at the start of file code (after imports) or at the top in the class.

File: blog/admin.py (hunk 8 above)
Comment (Ali-D-Akbar) [reply blog/admin.py]: I cannot change it as it is a built-in name. I'd add an in-line comment about what it is.

File: blog/admin.py (hunk 9 above)
Comment (Ali-D-Akbar) [reply blog/admin.py]: Imports have been sorted.
Added this to the makefile.

File: blog/admin.py (hunk 10 above)
Comment (Ali-D-Akbar) [reply blog/admin.py]: It's been removed.

File: blog/admin.py (hunk 11 above)
Comment (Ali-D-Akbar) [reply blog/admin.py]: Cannot directly display the date of birth because Profile is a one-to-one relation.

File: blog/models.py (hunk 12 above)
Comment (Ali-D-Akbar) [reply blog/models.py]: Done.

**Pull request merged**